### ISSUE TITLE
miou calculations are time-consuming

### ISSUE THREAD

**Issue by xiuguangLi** [Lightning-AI/torchmetrics#2891]
Hello, in the use of torchmetrics. Segmentation. MeanIoU calculation, more time-consuming, even with the cuda.
Example:
num_classes = 21
bs,h,w =10, 512,512
preds = torch.randint(0,num_classes,(bs,h,w))
labels = torch.randint(0,num_classes,(bs,h,w))

Can it be optimized?

**Comment by SkafteNicki** [comment]
Hi @xiuguangli, thanks for raising this issue.
I get somewhat the same results as @Isalia20 that the implementation is not really that slow. For completeness here i my benchmarking script
```python
import torch
from torchmetrics.segmentation import MeanIoU
import time
import pandas as pd
import numpy as np

def benchmark_miou(bs, device, repeats=10):
    num_classes = 21
    h, w = 512, 512
    preds = torch.randint(0, num_classes, (bs, h, w)).to(device)
    labels = torch.randint(0, num_classes, (bs, h, w)).to(device)

    miou = MeanIoU(num_classes=num_classes).to(device)
    
    times = []
    for _ in range(repeats):
        start_time = time.time()
        miou(preds, labels)
        end_time = time.time()
        times.append(end_time - start_time)
    
    total_time = sum(times)
    avg_time = np.mean(times)
    std_time = np.std(times)
    
    return total_time, avg_time, std_time

def run_benchmarks():
    devices = ['cpu', 'cuda'] if torch.cuda.is_available() else ['cpu']
    batch_sizes = [1, 4, 8, 16, 32, 64, 128]
    repeats = 10
    
    results = []
    
    for device in devices:
        for bs in batch_sizes:
            print(f"Benchmarking on {device} with batch size {bs}")
            total_time, avg_time, std_time = benchmark_miou(bs, device, repeats)
            results.append({
                'Device': device,
                'Batch Size': bs,
                'Total Time': total_time,
                'Average Time': avg_time,
                'Std Dev Time': std_time
            })
    
    return pd.DataFrame(results)

if __name__ == "__main__":
    results_df = run_benchmarks()
    print(results_df)
```
and the result from my laptop (GPU=Nvidia 4070)
```txt
   Device  Batch Size  Total Time  Average Time  Std Dev Time
0     cpu           1    0.004231      0.000423      0.000400
1     cpu           4    0.010362      0.001036      0.000402
2     cpu           8    0.082382      0.008238      0.000503
3     cpu          16    0.068503      0.006850      0.000128
4     cpu          32    0.130202      0.013020      0.000303
5     cpu          64    0.275238      0.027524      0.004126
6     cpu         128    0.494135      0.049413      0.001159
7    cuda           1    0.051750      0.005175      0.014914
8    cuda           4    0.002234      0.000223      0.000111
9    cuda           8    0.002421      0.000242      0.000149
10   cuda          16    0.002366      0.000237      0.000115
11   cuda          32    0.002348      0.000235      0.000108
12   cuda          64    0.002376      0.000238      0.000107
13   cuda         128    0.002514      0.000251      0.000137
```
overall my findings are what you usually find when comparing CPU to GPU: for very low batch sizes the CPU implementation is faster than the GPU implementation, which is usually due to the extra communication overhead between devices. However, for even for slightly larger batch sizes the extra communication between devices is offset by the high parallelization of computations on the GPU.
Closing issue as this is within the expectation of the performance that TM expects to deliver.
